File: packages/runtime-mcp/src/protocol.ts
// The subset of the daemon protocol this server speaks. Mirrors
// `crates/runtime-ipc/src/protocol.rs`; see `docs/architecture.md`.

export interface Project {
  id: string;
  name: string;
  root_path: string;
  repository_url?: string;
}

export interface Workspace {
  id: string;
  project_id: string;
  path: string;
  git_branch?: string;
  git_commit?: string;
  worktree: boolean;
  port_offset: number;
}

export type ServiceStatus =
  | "starting"
  | "healthy"
  | "unhealthy"
  | "stopping"
  | "stopped"
  | "failed"
  | "unknown";

export interface RuntimeInstance {
  id: string;
  service_id: string;
  pid: number;
  status: ServiceStatus;
  port?: number;
  started_at: string;
  exit_code?: number;
  started_by: string;
}

export interface ServiceView {
  id: string;
  workspace_id: string;
  name: string;
  service_type: string;
  command: string;
  cwd: string;
  preferred_port?: number;
  env?: Record<string, string>;
  status: ServiceStatus;
  instance?: RuntimeInstance;
  actual_port?: number;
  url?: string;
  /** False when the runtime found it already running and cannot stop it. */
  managed?: boolean;
  /** Services in the same checkout that must be up first. */
  depends_on?: string[];
  /** Runs to completion rather than staying up. */
  one_shot?: boolean;
  /** What a graceful stop sends, when SIGTERM is the wrong word for it. */
  stop_signal?: "term" | "int" | "quit" | "hup";
}

/** A live port in a checkout that no declared service explains. */
export interface ExternalService {
  port: number;
  pid: number;
  container?: string;
  cwd?: string;
  command_line?: string;
  url?: string;
}

/** A container compose defines for a checkout, running or not. */
export interface ContainerView {
  name: string;
  service?: string;
  image: string;
  status: string;
  health?: string;
  ports?: number[];
  url?: string;
}

export interface WorkspaceView extends Workspace {
  services: ServiceView[];
  external?: ExternalService[];
  containers?: ContainerView[];
}

export interface ProjectView extends Project {
  workspaces: WorkspaceView[];
  running_services: number;
  total_services: number;
  external_services?: number;
}

/** A project the runtime found on its own. */
export interface Discovery {
  root_path: string;
  name: string;
  running: boolean;
  ports?: number[];
  markers?: string[];
  git_branch?: string;
  suggested_services?: string[];
  registered: boolean;
}

export interface ProjectConfig {
  name?: string;
  services: Record<string, unknown>;
}

export interface PortOwner {
  port: number;
  /** One number can be held by both a TCP and a UDP socket. */
  protocol: "tcp" | "udp";
  pid: number;
  executable?: string;
  cwd?: string;
  command_line?: string;
  project_name?: string;
  git_branch?: string;
  service_name?: string;
  started_by?: string;
  /** Container publishing this port, when it is not a plain process. */
  container?: string;
  /** Another supervisor keeping this alive: "pm2", "systemd". */
  supervisor?: string;
  /** Only a process the runtime started may ever be terminated automatically. */
  managed: boolean;
}

/** A launch the runtime was told about but did not perform. */
export interface LaunchObservation {
  id: string;
  /** Exactly as given, never a script name inferred from it. */
  command: string;
  cwd: string;
  source: string;
  session?: string;
  observed_at: string;
  state: "pending" | "bound";
  port?: number;
  pid?: number;
  service_id?: string;
}

/** A service another supervisor keeps, that the runtime can switch. */
export interface SupervisedView {
  name: string;
  supervisor: string;
  status: string;
  pid?: number;
  command: string;
  restarts: number;
  /** Absent, not empty, when it holds none. */
  ports?: number[];
  url?: string;
  /** Set when restarting this would fail, with the reason. */
  restart_warning?: string;
}

/** A service that is not working, with the part of its output that says why. */
export interface Failure {
  service_id: string;
  /** `Loom/api`, as a person would say it. */
  subject: string;
  status: string;
  at: string;
  exit_code?: number;
  /** Absent, not empty, when it said nothing. */
  detail?: string[];
}

/** Something wrong with what is declared, found without being asked. */
export interface Finding {
  /** Where it is, as a person would say it: `Loom/api`. */
  subject: string;
  message: string;
  /** True when it will fail rather than merely might. */
  certain: boolean;
}

/** A named sequence of members in a checkout. */
export interface Stack {
  id: string;
  workspace_id: string;
  name: string;
  members: string[];
}

/** A stack with what its members are actually doing. */
export interface FlowNode {
  name: string;
  service_id?: string;
  after?: string[];
  /** How many waits deep it is. Everything on one level starts at once. */
  level: number;
  status: ServiceStatus;
  one_shot?: boolean;
}

export interface StackView extends Stack {
  services: ServiceView[];
  running: number;
  /** Members naming a service that no longer exists. */
  missing?: string[];
  /** The stack as a graph, worked out by the daemon from the members' own
   *  dependencies so every surface reads one shape. */
  flow?: FlowNode[];
}

export interface AdoptOutcome {
  /** The stack it was put in, so it can be started afterwards. */
  stack?: string;
  service: ServiceView;
  /** Where the command came from. Never the project's scripts. */
  command_source: "recorded" | "process_argv" | "supervisor";
  /** False when the service was already declared and nothing changed. */
  declared: boolean;
  /** The command written down before, when adopting replaced it. */
  replaced_command?: string;
  supervisor?: string;
}

export interface PortStatus {
  port: number;
  available: boolean;
  owner?: PortOwner;
  suggested_port?: number;
}

export interface PortReservation {
  port: number;
  preferred_port?: number;
  reallocated: boolean;
  policy: string;
  conflict?: PortOwner;
}

export interface HealthReport {
  service_id: string;
  status: ServiceStatus;
  detail?: string;
  checked_port?: number;
}

export interface LogLine {
  seq: number;
  service_id: string;
  stream: "stdout" | "stderr" | "system";
  timestamp: string;
  message: string;
}

export interface StartOutcome {
  service: ServiceView;
  reused: boolean;
  reservation?: PortReservation;
  /** Something about this start worth knowing first. */
  warning?: string;
}

export interface AgentSession {
  id: string;
  provider: string;
  client: string;
}

export interface DaemonInfo {
  version: string;
  pid: number;
  platform: string;
  socket_path: string;
  database_path: string;
  uptime_seconds: number;
}

/** A request as it appears on the wire: `{ method, params }`. */
export interface Request {
  method: string;
  params?: Record<string, unknown>;
}

export type ResponseBody =
  | { type: "pong"; protocol_version: number }
  | ({ type: "info" } & DaemonInfo)
  | { type: "projects"; items: ProjectView[] }
  | { type: "discoveries"; items: Discovery[] }
  | ({ type: "project" } & ProjectView)
  | { type: "workspaces"; items: Workspace[] }
  | ({ type: "workspace" } & Workspace)
  | { type: "services"; items: ServiceView[] }
  | ({ type: "config" } & ProjectConfig)
  | ({ type: "container" } & ContainerView)
  | ({ type: "service" } & ServiceView)
  | ({ type: "started" } & StartOutcome)
  | ({ type: "health" } & HealthReport)
  | ({ type: "port" } & PortStatus)
  | { type: "ports"; items: PortOwner[] }
  | ({ type: "reservation" } & PortReservation)
  | { type: "logs"; items: LogLine[] }
  | { type: "launches"; items: LaunchObservation[] }
  | ({ type: "supervised" } & SupervisedView)
  | { type: "findings"; items: Finding[] }
  | { type: "failures"; items: Failure[] }
  | { type: "stacks"; items: StackView[] }
  | { type: "stack_run"; done: string[] }
  | ({ type: "adopted" } & AdoptOutcome)
  | { type: "sessions"; items: AgentSession[] }
  | ({ type: "session" } & AgentSession)
  | { type: "done"; ok: boolean };

export interface RuntimeErrorPayload {
  code: string;
  [key: string]: unknown;
}

export type Frame =
  | { kind: "request"; id: number; request: Request }
  | { kind: "response"; id: number; result: ResponseBody }
  | { kind: "error"; id: number; error: RuntimeErrorPayload; message: string }
  | { kind: "event"; event: Record<string, unknown> };

File: crates/runtime-ipc/tests/protocol.rs
//! Every frame must survive the wire.
//!
//! Serde's internally tagged enums cannot encode a newtype variant wrapping a
//! sequence or a string — and the failure is at *runtime*, while writing the
//! response, so the daemon drops the connection having said nothing. That has
//! happened twice: once to `ResponseBody`'s collections, once to
//! `RuntimeError`'s messages. This test exists so it cannot happen a third time
//! without a red test.

use std::collections::BTreeMap;

use chrono::Utc;
use runtime_core::discover::Discovery;
use runtime_ipc::protocol::{Frame, Request, ResponseBody};
use runtime_types::*;

fn service() -> Service {
    Service {
        id: ServiceId::from("svc"),
        workspace_id: WorkspaceId::from("ws"),
        name: "web".to_string(),
        service_type: ServiceType::Web,
        command: "pnpm dev".to_string(),
        cwd: "/repo".into(),
        env: BTreeMap::from([("A".to_string(), "1".to_string())]),
        preferred_port: Some(3000),
        health_check: Some(HealthCheck::Tcp { port: None }),
        auto_start: false,
        conflict_policy: ConflictPolicy::AllocateNext,
        depends_on: vec!["db".to_string()],
        one_shot: false,
    }
}

fn workspace() -> Workspace {
    Workspace {
        id: WorkspaceId::from("ws"),
        project_id: ProjectId::from("proj"),
        path: "/repo".into(),
        git_branch: Some("main".to_string()),
        git_commit: None,
        worktree: false,
        port_offset: 0,
        created_at: Utc::now(),
    }
}

fn service_view() -> ServiceView {
    ServiceView {
        service: service(),
        status: ServiceStatus::Healthy,
        instance: None,
        actual_port: Some(3000),
        url: Some("http://localhost:3000".to_string()),
        managed: true,
        supervisor: Some("pm2".to_string()),
        supervisor_entry: Some("flip7".to_string()),
    }
}

fn project_view() -> ProjectView {
    ProjectView {
        project: Project {
            id: ProjectId::from("proj"),
            name: "shop".to_string(),
            root_path: "/repo".into(),
            repository_url: None,
            created_at: Utc::now(),
            updated_at: Utc::now(),
        },
        workspaces: vec![WorkspaceView {
            workspace: workspace(),
            services: vec![service_view()],
            external: vec![ExternalService {
                port: 5555,
                pid: 1,
                container: None,
                cwd: None,
                command_line: Some("node".to_string()),
                url: None,
                supervisor: Some("pm2".to_string()),
            }],
            supervised: vec![],
            containers: vec![ContainerView {
                name: "db".to_string(),
                service: Some("db".to_string()),
                image: "postgres:16".to_string(),
                status: "running".to_string(),
                health: Some("healthy".to_string()),
                ports: vec![5432],
                url: None,
            }],
        }],
        running_services: 1,
        total_services: 1,
        external_services: 1,
    }
}

fn port_owner() -> PortOwner {
    PortOwner {
        port: 3000,
        pid: 42,
        executable: Some("/usr/local/bin/node".to_string()),
        cwd: Some("/repo".into()),
        command_line: Some("node server.js".to_string()),
        project_id: Some(ProjectId::from("proj")),
        project_name: Some("shop".to_string()),
        workspace_id: Some(WorkspaceId::from("ws")),
        git_branch: Some("main".to_string()),
        service_id: Some(ServiceId::from("svc")),
        service_name: Some("web".to_string()),
        started_by: Some(StartedBy::Cli),
        container: None,
        supervisor: Some("pm2".to_string()),
        managed: true,
    }
}

/// One of every `ResponseBody` variant.
fn responses() -> Vec<ResponseBody> {
    vec![
        ResponseBody::Pong { protocol_version: 1 },
        ResponseBody::Info(DaemonInfo {
            version: "0.1.0".to_string(),
            pid: 1,
            socket_path: "/tmp/s.sock".into(),
            database_path: "/tmp/db".into(),
            platform: "macos".to_string(),
            uptime_seconds: 3,
        }),
        ResponseBody::Projects { items: vec![project_view()] },
        ResponseBody::Project(project_view()),
        ResponseBody::Discoveries {
            items: vec![Discovery {
                root_path: "/repo".into(),
                name: "shop".to_string(),
                running: true,
                ports: vec![3000],
                markers: vec!["git".to_string()],
                git_branch: Some("main".to_string()),
                suggested_services: vec!["web".to_string()],
                registered: false,
            }],
        },
        ResponseBody::Workspaces { items: vec![workspace()] },
        ResponseBody::Workspace(workspace()),
        ResponseBody::Services { items: vec![service_view()] },
        ResponseBody::Service(service_view()),
        ResponseBody::Config(ProjectConfig {
            name: Some("shop".to_string()),
            services: BTreeMap::from([(
                "web".to_string(),
                ServiceConfig {
                    command: "pnpm dev".to_string(),
                    port: Some(3000),
                    cwd: None,
                    service_type: Some(ServiceType::Web),
                    env: BTreeMap::new(),
                    health: None,
                    auto_start: false,
                    on_conflict: None,
                    depends_on: vec!["db".to_string()],
                    one_shot: true,
                },
            )]),
        }),
        ResponseBody::Container(ContainerView {
            name: "db".to_string(),
            service: None,
            image: "postgres:16".to_string(),
            status: "exited".to_string(),
            health: None,
            ports: vec![],
            url: None,
        }),
        ResponseBody::Started(StartOutcome {
            warning: Some("rewrites a build 'flip7' serves from".to_string()),
            service: service_view(),
            reused: false,
            reservation: Some(PortReservation {
                port: 3000,
                preferred_port: Some(3000),
                reallocated: false,
                policy: ConflictPolicy::Fail,
                conflict: Some(port_owner()),
            }),
        }),
        ResponseBody::Health(HealthReport {
            service_id: ServiceId::from("svc"),
            status: ServiceStatus::Unhealthy,
            detail: Some("refused".to_string()),
            checked_port: Some(3000),
        }),
        ResponseBody::Port(PortStatus {
            port: 3000,
            available: false,
            owner: Some(port_owner()),
            lease_status: Some(PortLeaseStatus::Active),
            suggested_port: Some(3001),
        }),
        ResponseBody::Ports { items: vec![port_owner()] },
        ResponseBody::Reservation(PortReservation {
            port: 3001,
            preferred_port: Some(3000),
            reallocated: true,
            policy: ConflictPolicy::AllocateNext,
            conflict: None,
        }),
        ResponseBody::Logs {
            items: vec![LogLine {
                seq: 0,
                service_id: ServiceId::from("svc"),
                stream: LogStream::Stderr,
                timestamp: Utc::now(),
                message: "boom".to_string(),
            }],
        },
        ResponseBody::Setting { value: Some("{}".to_string()) },
        ResponseBody::Launches {
            items: vec![LaunchObservation {
                id: "1-abc".to_string(),
                command: "cd frontend && PORT=4000 pnpm dev".to_string(),
                cwd: "/repo".into(),
                source: StartedBy::ClaudeCode,
                session: Some("s1".to_string()),
                observed_at: Utc::now(),
                state: LaunchState::Bound,
                port: Some(4000),
                pid: Some(42),
                service_id: Some(ServiceId::from("svc")),
            }],
        },
        ResponseBody::Adopted(AdoptOutcome {
            service: service_view(),
            command_source: CommandSource::ProcessArgv,
            declared: false,
            replaced_command: Some("npm run dev".to_string()),
            supervisor: Some("pm2".to_string()),
        }),
        ResponseBody::Supervised(SupervisedView {
            name: "flip7".to_string(),
            supervisor: "pm2".to_string(),
            status: "online".to_string(),
            pid: Some(29106),
            command: "server.mjs".to_string(),
            restarts: 22,
            ports: vec![3007],
            url: Some("http://localhost:3007".to_string()),
            restart_warning: Some("holds a development build".to_string()),
        }),
        ResponseBody::Tasks {
            items: vec![Task {
                id: TaskId::from("task"),
                workspace_id: WorkspaceId::from("ws"),
                name: "dev".to_string(),
                steps: vec!["migrate".to_string(), "api".to_string()],
            }],
        },
        ResponseBody::TaskRun {
            steps: vec!["migrate (ran)".to_string(), "api".to_string()],
        },
        ResponseBody::Sessions {
            items: vec![AgentSession {
                id: SessionId::from("sess"),
                provider: "anthropic".to_string(),
                client: "claude-code".to_string(),
                cwd: None,
                project_id: None,
                started_at: Utc::now(),
                last_seen_at: Utc::now(),
            }],
        },
        ResponseBody::Session(AgentSession {
            id: SessionId::from("sess"),
            provider: "openai".to_string(),
            client: "codex".to_string(),
            cwd: Some("/repo".into()),
            project_id: Some(ProjectId::from("proj")),
            started_at: Utc::now(),
            last_seen_at: Utc::now(),
        }),
        ResponseBody::Done { ok: true },
    ]
}

#[test]
fn every_response_survives_a_round_trip() {
    for body in responses() {
        let label = format!("{body:?}");
        let frame = Frame::Response { id: 1, result: body };

        let encoded = serde_json::to_string(&frame)
            .unwrap_or_else(|err| panic!("cannot encode {label}: {err}"));
        let decoded: Frame = serde_json::from_str(&encoded)
            .unwrap_or_else(|err| panic!("cannot decode {label}: {err}\n{encoded}"));
        assert_eq!(decoded, frame, "changed shape: {label}");
    }
}

#[test]
fn every_request_survives_a_round_trip() {
    let requests = vec![
        Request::Ping,
        Request::DaemonInfo,
        Request::Shutdown,
        Request::ListProjects,
        Request::DiscoverProjects { paths: vec!["/a".into()], adopt: true },
        Request::GetProject { selector: "shop".to_string() },
        Request::AddProject { path: "/repo".into(), name: Some("shop".to_string()) },
        Request::RemoveProject { selector: "shop".to_string() },
        Request::ListWorktrees { selector: "shop".to_string() },
        Request::RegisterWorktree { selector: "shop".to_string(), path: "/wt".into() },
        Request::ListServices { project: None },
        Request::GetService { project: None, service: "web".to_string() },
        Request::UpdateService {
            project: None,
            service: "web".to_string(),
            patch: ServicePatch {
                preferred_port: Some(None),
                env: BTreeMap::from([("A".to_string(), "1".to_string())]),
                remove_env: vec!["B".to_string()],
                ..Default::default()
            },
        },
        Request::AddService {
            selector: "shop".to_string(),
            name: "worker".to_string(),
            config: ServiceConfig {
                command: "pnpm worker".to_string(),
                port: None,
                cwd: None,
                service_type: None,
                env: BTreeMap::new(),
                health: None,
                auto_start: false,
                on_conflict: None,
                depends_on: vec!["db".to_string()],
                one_shot: true,
            },
        },
        Request::RemoveService { project: None, service: "web".to_string() },
        Request::ExportConfig { selector: "shop".to_string() },
        Request::RecordLaunch {
            command: "cd frontend && PORT=4000 pnpm dev".to_string(),
            cwd: "/repo".into(),
            source: Some("claude-code".to_string()),
            session: Some("s1".to_string()),
        },
        Request::ListLaunches,
        Request::ListTasks { selector: ".".to_string() },
        Request::SetTask {
            selector: ".".to_string(),
            name: "dev".to_string(),
            steps: vec!["migrate".to_string(), "api".to_string()],
        },
        Request::RemoveTask { selector: ".".to_string(), name: "dev".to_string() },
        Request::RunTask { selector: ".".to_string(), name: "dev".to_string() },
        Request::ControlSupervised {
            name: "flip7".to_string(),
            action: "restart".to_string(),
        },
        Request::AdoptPort { port: 3007, force: true },
        Request::StartService {
            project: None,
            service: "web".to_string(),
            port: Some(3000),
            on_conflict: Some("fail".to_string()),
            started_by: Some("cli".to_string()),
            session: None,
        },
        Request::StopService { project: None, service: "web".to_string(), timeout_seconds: Some(5) },
        Request::RestartService {
            project: None,
            service: "web".to_string(),
            started_by: None,
            session: None,
        },
        Request::ControlContainer { name: "db".to_string(), action: "stop".to_string() },
        Request::GetContainerLogs { name: "db".to_string(), max_lines: Some(10) },
        Request::GetHealth { project: None, service: "web".to_string() },
        Request::WaitUntilHealthy { project: None, service: "web".to_string(), timeout_seconds: None },
        Request::CheckPort { port: 3000 },
        Request::ListPorts,
        Request::ReservePort {
            project: None,
            service: "web".to_string(),
            port: None,
            on_conflict: None,
            started_by: None,
        },
        Request::ReleasePort { port: 3000 },
        Request::GetLogs {
            project: None,
            service: "web".to_string(),
            max_lines: Some(50),
            since_seq: Some(3),
        },
        Request::GetSetting { key: "desktop.panel".to_string() },
        Request::SetSetting { key: "k".to_string(), value: "v".to_string() },
        Request::ListSessions,
        Request::RegisterSession {
            provider: "anthropic".to_string(),
            client: "claude-code".to_string(),
            cwd: None,
        },
        Request::Subscribe,
    ];

    for request in requests {
        let label = format!("{request:?}");
        let frame = Frame::Request { id: 1, request };

        let encoded = serde_json::to_string(&frame)
            .unwrap_or_else(|err| panic!("cannot encode {label}: {err}"));
        let decoded: Frame = serde_json::from_str(&encoded)
            .unwrap_or_else(|err| panic!("cannot decode {label}: {err}\n{encoded}"));
        assert_eq!(decoded, frame, "changed shape: {label}");
    }
}
